### ISSUE TITLE
Linked backends: Streaming does not stream; it arrives in a single payload

### ISSUE THREAD

**Issue by johnnyreilly** [Azure/static-web-apps#1180]
**Describe the bug**

Imagine an Azure Static Web App with a linked backend.  The linked backend is an Azure Function App, but based upon our investigations; the issue does not appear to be function app related.

**To Reproduce**

Deploy a Static Web App and a Function App which is a linked backend to the SWA.  The backend contains this streaming function named `GetChatCompletionsStream`: 

```cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Azure.WebJobs;
using Microsoft.Azure.WebJobs.Extensions.Http;
using Microsoft.Azure.WebJobs.Extensions.OpenApi.Core.Attributes;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.Logging;
using System.Net;
using System.Text.Json;
using Azure;
using Azure.AI.OpenAI;
using System.Linq;

namespace ZebraGptFunctionApp.Functions;

public record CompletePromptParameters(
    string prompt,
    double temperature,
    int maxTokens,
    string deploymentName,
    bool canRecord
);

public class OpenAiFunction
{
    private readonly AppSettings _appSettings;
    private readonly OpenAIClient _openAIClient;
    private readonly ILogger<OpenAiFunction> _log;

    public OpenAiFunction(
        AppSettings appSettings,
        ILogger<OpenAiFunction> log,
        OpenAIClient openAIClient
    )
    {
        _appSettings = appSettings;
        _log = log;
        _openAIClient = openAIClient;
    }

    public record ChatCompletionParameters(
        string DeploymentName,
        float Temperature,
        int MaxTokens,
        List<ChatMessageParameter> Messages,
        bool CanRecord
    );

    public record ChatMessageParameter(string Role, string Content);

    [FunctionName(nameof(OpenAiFunction.GetChatCompletionsStream))]
    public async Task<IActionResult> GetChatCompletionsStream(
        [HttpTrigger(AuthorizationLevel.Anonymous)] HttpRequest req
    )
    {
        try
        {
            var chatCompletionParameters = await GetChatCompletionParameters(req);
            var chatOptions = GetChatCompletionsOptions(chatCompletionParameters);

            var response = req.HttpContext.Response;

            response.StatusCode = (int)HttpStatusCode.OK;
            // response.ContentType = "text/plain";
            response.ContentType = "text/event-stream";

            await using var sw = new StreamWriter(response.Body);

            var streamingChatCompletionsResponse = await _openAIClient.Client.GetChatCompletionsStreamingAsync(
                deploymentOrModelName: chatCompletionParameters.DeploymentName,
                chatOptions
            );

            using StreamingChatCompletions streamingChatCompletions = streamingChatCompletionsResponse.Value;

            await foreach (StreamingChatChoice choice in streamingChatCompletions.GetChoicesStreaming())
            {
                await foreach (ChatMessage message in choice.GetMessageStreaming())
                {
                    // THIS IS STREAMING
                    await sw.WriteAsync(message.Content);
                    await sw.FlushAsync();
                }
            }
        }
        catch (Exception ex)
        {
            _log.LogError(ex, "Problem getting chat completion");
            req.HttpContext.Response.StatusCode = (int)HttpStatusCode.InternalServerError;
        }

        // Required to avoid the host trying to add headers to the response
        return new EmptyResult();
    }

    private async static Task<ChatCompletionParameters> GetChatCompletionParameters(HttpRequest req)
    {
        var content = await new StreamReader(req.Body).ReadToEndAsync();

        var chatCompletionParameters =
            JsonSerializer.Deserialize<ChatCompletionParameters>(content, new JsonSerializerOptions
            {
                PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
            })!;

        return chatCompletionParameters;
    }

    private static ChatCompletionsOptions GetChatCompletionsOptions(ChatCompletionParameters chatCompletionParameters)
    {
        var chatRoleMapping = new Dictionary<string, ChatRole>()
            {
                { "system", ChatRole.System }, { "user", ChatRole.User }, { "assistant", ChatRole.Assistant }
            };

        var chatOptions = new ChatCompletionsOptions()
        {
            Temperature = chatCompletionParameters.Temperature,
            MaxTokens = chatCompletionParameters.MaxTokens,
            NucleusSamplingFactor = (float)0.95,
            FrequencyPenalty = 0,
            PresencePenalty = 0,
        };

        // it has to be added like this because by design Messages is a readonly. https://github.com/Azure/azure-sdk-for-net/issues/35096
        chatCompletionParameters.Messages.ForEach(message => 
            chatOptions.Messages.Add(new ChatMessage(chatRoleMapping[message.Role], message.Content)));

        return chatOptions;
    }
}
```

Eagle peeps will note we're building an Open AI chat mechanism - but what's significant here is we stream text to the caller.

On the front end we have a TypeScript function that looks like this:

```ts
        try {
            if (!userInput.trim()) {
                return;
            }

            const userMessage: ChatMessage = {
                role: "user",
                content: userInput,
            };
            setMessages((prevMessages) => [...prevMessages, userMessage]);
            setUserInput("");
            setIsLoading(true);

            const response = await fetch("/api/GetChatCompletionsStream", {
                method: "POST",
                body: JSON.stringify({
                    deploymentName: "OpenAi-gpt-35-turbo",
                    maxTokens,
                    temperature,
                    messages: [...messages, userMessage],
                } as ChatCompletionParameters),
                headers: {
                    Accept: "text/event-stream",
                    "Content-Type": "application/json",
                },
            });

            if (!response.body) {
                return;
            }

            const reader = response.body.getReader();

            let reads = 0;
            let responseMessage = "";
            // eslint-disable-next-line no-constant-condition
            while (true) {
                const { done, value } = await reader.read();

                if (done) {
                    break;
                }

                const fragment = new TextDecoder("utf-8").decode(value);
                console.log(`fragment #${++reads}`, fragment);
                responseMessage += fragment;
                setLiveResponseMessage(responseMessage);
            }
            console.log(`done in ${reads} reads`);
            setMessages((prevMessages) => [
                ...prevMessages,
                { role: "assistant", content: responseMessage },
            ]);
            setIsLoading(false);
            setLiveResponseMessage("");
        } catch (error) {
            console.error("Error sending message:", error);
        }
```

**Expected behavior**

Running locally, this works as expected: streaming.  Deployed - it does not; we get a *single* payload in our "stream". 

**Screenshots**

**Additional context**

**Comment by thomasgauvin** [comment]
Thanks for reporting @johnnyreilly, let me bring this up to eng team

**Comment by thomasgauvin** [comment]
I've brought this up to our eng team and it's on our backlog

**Comment by johnnyreilly** [comment]
Thanks @thomasgauvin!

**Comment by ChouchouCendre** [comment]
Same issue here, new stream.PassThrough() is working correctly locally but it's not working on the server.
Everything arrives at the end.

@thomasgauvin any idea when it'll fixed? Are we talking days, weeks?

**Comment by thomasgauvin** [comment]
We tried reproducing this but it seems we're able to get streams working. For your linked backend, are you using in-process or isolated for the .NET runtime? https://learn.microsoft.com/en-us/azure/azure-functions/dotnet-isolated-in-process-differences

**Comment by johnnyreilly** [comment]
Hi @thomasgauvin,

We're using .NET 6 in process Azure Functions. I'd be happy to jump on a call and demo if you'd like.

**Comment by thomasgauvin** [comment]
Ok thanks for the details @johnnyreilly, I think we were able to repro by adding the call to OpenAI and attempting to stream the response back. We're continuing to investigate

**Comment by v1212** [comment]
@johnnyreilly, thanks for the detail, we are working on a fix in SWA to return the stream response from function app API, before the fix rolling out in SWA product, a workaround to this issue can be calling direct API of function app instead of that of SWA API path, that worked in our investigation, that need an extra CORS setting in the target function app allowing call from web page of SWA app.

**Comment by johnnyreilly** [comment]
Thanks @v1212 - I'm aware that it's possible to handle calling the function directly; but we struggled to get the Azure AD authentication working with our front end code. In the end we pivoted to using Azure Container Apps with Easy Auth. Great to hear a fix is on the way though!

**Comment by thomasgauvin** [comment]
Update: We've been investigating this and determined that this depends on our underlying infrastructure. We're working with our partner teams to see how we could support this.

**Comment by pietz** [comment]
> @johnnyreilly, thanks for the detail, we are working on a fix in SWA to return the stream response from function app API, before the fix rolling out in SWA product, a workaround to this issue can be calling direct API of function app instead of that of SWA API path, that worked in our investigation, that need an extra CORS setting in the target function app allowing call from web page of SWA app.

Could you elaborate on that? I'm not able to setup streaming in an Azure Function (Python) in any way. Are you suggesting to separate frontend and backend, in order to have a separate Azure Function? Because I tried that and it didn't work.

**Comment by martgra** [comment]
Any update here?

Experiencing same - streaming is not working between Static Web app linked to Web app backend.
Frontend: react
Backend: docker container + fastapi. Can provide more details on request.

**Comment by nkma1989** [comment]
Hi,
I have this exact issue, we are using Azure Web App to host a docker container with a Fast API, talking to Azure OpenAI service, using streaming responses. 
When I run locally it's streaming and everything works as expected, but when deploying to azure web app it does not stream, but waits for the full response and then returns.
We are using python 3.10 and a gunicorn server to host the app.
@thomasgauvin is this also something that you are looking at?

**Comment by philipmuller** [comment]
Hello!

Experiencing the same issue when using nextjs with the [Vercel AI SDK](https://sdk.vercel.ai/docs/introduction) deployed on Azure SWA. Any updates or workarounds?

**Comment by kunal1054** [comment]
We are also facing same issue with angular front end and .net core 7 backend deployed on azure app service. Response is returned as a single response. Works locally. We can directly call open ai streaming chat endpoint as that would expose the api key. This has to be resolved asap

**Comment by egerszu** [comment]
Same error. We linked to Azure Functions, and SWA - Azure Function does not stream. Local - Azure Functions works perfectly.

**Comment by simplecohesion** [comment]
Same here (nextjs + vercel ai sdk).  Is there any hope of this being supported, as the thread is over a year old?

**Comment by hectordanielc** [comment]
Same here (nextjs + vercel ai sdk).

**Comment by lukju** [comment]
I've built both, a Python and and a Node/Javascript API. In both cases, streaming does work locally but not as a deployed SWA. Any ideas or workarounds?

**Comment by egerszu** [comment]
> I've built both, a Python and and a Node/Javascript API. In both cases, streaming does work locally but not as a deployed SWA. Any ideas or workarounds?

The only workaround is that instead of using /api/, use https://your-azure-func.azurewebsites.net/api/whatever/?code=your-function-key

Streaming works this way.

**Comment by hectordanielc** [comment]
> > I've built both, a Python and and a Node/Javascript API. In both cases, streaming does work locally but not as a deployed SWA. Any ideas or workarounds?
> 
> The only workaround is that instead of using /api/, use https://your-azure-func.azurewebsites.net/api/whatever/?code=your-function-key
> 
> Streaming works this way.

Still not streaming for me.

**Comment by thomasgauvin** [comment]
(Disclaimer, no longer PM for SWA) I wrote https://techcommunity.microsoft.com/t5/apps-on-azure-blog/add-a-context-grounded-ai-chatbot-to-your-azure-static-web-apps/ba-p/4097223#:~:text=Create%20the%20API%20for%20our%20chatbot during my last couple of weeks on the SWA team, might be helpful? Note the `text/event-stream` header used in the response from the Azure Function. I don't know the current status of SWA anymore but I think that blog should be a helpful resource. Keep me updated!

**Comment by hectordanielc** [comment]
> (Disclaimer, no longer PM for SWA) I wrote https://techcommunity.microsoft.com/t5/apps-on-azure-blog/add-a-context-grounded-ai-chatbot-to-your-azure-static-web-apps/ba-p/4097223#:~:text=Create%20the%20API%20for%20our%20chatbot during my last couple of weeks on the SWA team, might be helpful? Note the `text/event-stream` header used in the response from the Azure Function. I don't know the current status of SWA anymore but I think that blog should be a helpful resource. Keep me updated!

It didn't work for me but my azure function is made in python. I'll try it with node.

**Comment by lukju** [comment]
> > (Disclaimer, no longer PM for SWA) I wrote https://techcommunity.microsoft.com/t5/apps-on-azure-blog/add-a-context-grounded-ai-chatbot-to-your-azure-static-web-apps/ba-p/4097223#:~:text=Create%20the%20API%20for%20our%20chatbot during my last couple of weeks on the SWA team, might be helpful? Note the `text/event-stream` header used in the response from the Azure Function. I don't know the current status of SWA anymore but I think that blog should be a helpful resource. Keep me updated!
> 
> It didn't work for me but my azure function is made in python. I'll try it with node.

No need to go that extra route: the programming language doesn't matter - streaming doesn't work for all flavors of SWA APIs. 
The only workaround is to deploy the API as a dedicated Functions app and call those endpoint directly (do not link it with the SWA).

**Comment by lukju** [comment]
Does anybody know if Microsoft is aware of this problem?
As response streaming definitely does not work in Static Web Apps, it lowers the usefullness of this service in the context of ChatBots nearly to Zero.  Would be great if this could somehow be fixed.
Please please, you great makers of this otherwise great service: Let's make it streaming-able :-)

**Comment by v1212** [comment]
@lukju could you set content type in the api response to "application/octet-stream", or "text/event-stream"? these are two types currently Static Web App detects as streaming content, and also the mostly common types for AI streaming response.

**Comment by lukju** [comment]
What kind of code do you need? Did you see the code posted by @egerszu last Friday or so? It pretty much shows what needs to be done

**Comment by hectordanielc** [comment]
Just to see how are you constructing the endpoint. Mine is not working with media_type="application/octet-stream".

**Comment by lukju** [comment]
example in python:
![image](https://github.com/user-attachments/assets/1b438675-374e-40ca-bc9e-0099118c96ce)

example in typescript:
![image](https://github.com/user-attachments/assets/600cefc5-9213-4a83-8f6d-d6612cb37ecd)

**Comment by TheoMcCabe** [comment]
Same i have this issue using nuxt on an azure static web app . app works fine locally but streaming binary to my backend when deploying to SWA fails

**Comment by TheoMcCabe** [comment]
This is mad that this still doesnt work in the context of everyone building AI apps which make use of streaming

**Comment by v1212** [comment]
steaming with type "text/event-stream" and "application/octet-stream" works in Static Web App, "text/event-stream" is commonly used for text-based event streaming, and "application/octet-stream" is for binary data transmitting.

such as an API with type "text/event-stream"
![Image](https://github.com/user-attachments/assets/7cc822ec-dee0-47ba-9ed2-7a183d12ddfd)

the API can be accessed directly from Azure function app https://[functionapp2108242.azurewebsites.net/api/function3](https://functionapp2108242.azurewebsites.net/api/function3) and also available via Static Web App https://[ashy-ocean-02dbac200.5.azurestaticapps.net/api/function3](https://ashy-ocean-02dbac200.5.azurestaticapps.net/api/function3), the behaviors are the same as streaming.

If the client browser needs to handle each read of the streaming, a quick javascript as below works. 

    <script>
        async function sendPost() {
            // Get the input value
            var inputValue = document.getElementById('myAsk').value;

            // The URL of the API you want to call
            var url = "/";

            // Call the api in the site
            const data = { Question: inputValue };
            try {
                const response = await fetch(url + "api/myFunctionAPI", {
                    method: 'POST',
                    headers: {
                        'Content-Type': 'application/json'
                    },
                    body: JSON.stringify(data)
                });

                if (!response.body) {
                    console.log("response body is null");
                }
                else {
                    // proceed each read from the steaming response
                    var reader = response.body.getReader();
                    var reads = 0;
                    var content = "";
                    document.getElementById('responseBox').value = "";
                    document.getElementById('myCode').innerText = "";
                    // eslint-disable-next-line no-constant-condition
                    while (true) {
                        console.log("proceed ".concat(reads++, " reads"));
                        var _a = await reader.read(), done = _a.done, value = _a.value;
                        
                        if (done) {
                            var fragment_1 = new TextDecoder("utf-8").decode(value);
                            content += fragment_1;
                            document.getElementById('responseBox').value += fragment_1;
                            document.getElementById('dynamicCode').innerHTML = content;
                            Prism.highlightAll();

                            break;
                        }
                        var fragment = new TextDecoder("utf-8").decode(value);
                        content += fragment;
                        document.getElementById('responseBox').value += fragment;
                        document.getElementById('dynamicCode').innerHTML = content;
                    }
                    Prism.highlightAll();
                }
            }
            catch (error) {
                console.error("Error happened:", error);
            }
        }
    </script>

**Comment by educastroa** [comment]
> application/octet-stream
changing from text/event-stream to application/octet-stream worked for me as well

**Comment by thezanke** [comment]
For those here running into this using FastAPI, I managed to get it to work. Even after reading this thread and some others, it wasn't obvious what the fix was right away since it's caused by something happening under the hood.

### TL;DR
You need to add `headers={ "Content-Type":  "text/event-stream" }` to your `StreamingResponse` call.

### Explanation
It seems like the SWA platform is looking for EXACTLY "Content-Type: text/event-stream" to enable streaming functionality to work. Using Starlette/FastAPI's `StreamingResponse` with `media_type="text/event-stream"` works in theory, but Starlette is adding a charset property to the content-type making it `"text/even-stream; charset: utf-8"` which is then not correctly matching the expected content-type from the SWA platform.

Once you add the headers section explicitly, it appears to correct the issues with streaming.

**Comment by p-krukenfellner** [comment]
> You need to add `headers={ "Content-Type": "text/event-stream" }` to your `StreamingResponse` call.

I can confirm that this works in a **FastAPI** environment.

@thezanke Thank's a lot for this subtle but very effective change!
